### PR TITLE
CDAP-1899: batch writing metric values to MetricStore

### PR DESCRIPTION
**Pull request by abaranau**
https://issues.cask.co/browse/CDAP-1899

### CODE REVIEW

**Comment by abaranau** [comment]
http://builds.cask.co/browse/CDAP-RBT205-1

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/timeseries/FactTable.java [hunk 1]
```diff
@@ -116,6 +116,7 @@ public void add(List<Fact> facts) throws Exception {
     NavigableMap<byte[], NavigableMap<byte[], Long>> convertedGaugesTable =
       Maps.transformValues(gaugesTable, TRANSFORM_MAP_BYTE_ARRAY_TO_LONG);
 
+    // todo: replace with single call, to be able to optimize rpcs in underlying table
```
Comment (gsps1) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/timeseries/FactTable.java]: you mean the increment and put into one call ? description is not clear.

Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/timeseries/FactTable.java]: yes, the below two lines

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/cube/DefaultCube.java [hunk 2]
```diff
@@ -67,19 +68,26 @@ public DefaultCube(int[] resolutions, FactTableSupplier factTableSupplier,
...
-      if (agg.accept(fact)) {
-        List<TagValue> tagValues = Lists.newArrayList();
-        for (String tagName : agg.getTagNames()) {
-          tagValues.add(new TagValue(tagName, fact.getTagValues().get(tagName)));
+    add(ImmutableList.of(fact));
+  }
+
+  @Override
+  public void add(Collection<CubeFact> facts) throws Exception {
+    List<Fact> toWrite = Lists.newArrayList();
```
Comment (gsps1) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/cube/DefaultCube.java]: nit : matchingFacts or aggregateFacts ?

Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/cube/DefaultCube.java]: yeah - spent time picking the right name. I think the one I used is least confusing though :)

**Comment by gsps1** [comment]
LGTM  :+1:

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/cube/Cube.java [hunk 3]
```diff
@@ -35,6 +35,12 @@
   void add(CubeFact fact) throws Exception;
 
   /**
+   * Adds {@link CubeFact}s to this {@link Cube}.
+   * @param facts facts to add.
+   */
+  void add(Collection<CubeFact> facts) throws Exception;
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/cube/Cube.java]: Should be `Collection<? extends CubeFact>`

**Comment by abaranau** [comment]
also confirmed on perf cluster - we saw processing was behind, with this patch - it keeps up

File: cdap-watchdog-api/src/main/java/co/cask/cdap/api/metrics/MetricStore.java [hunk 4]
```diff
@@ -31,6 +31,13 @@
   void add(MetricValue metricValue) throws Exception;
 
   /**
+   * Adds {@link MetricValue}s to the store.
+   * @param metricValues metric values to add.
+   * @throws Exception
+   */
+  void add(Collection<MetricValue> metricValues) throws Exception;
```
Comment (chtyim) [inline cdap-watchdog-api/src/main/java/co/cask/cdap/api/metrics/MetricStore.java]: `Collection<? extends MetricValue>`

**Comment by sreevatsanraman** [comment]
http://builds.cask.co/browse/CDAP-RBT205-3

**Comment by albertshau** [comment]
minor nit, otherwise lgtm

**Comment by sreevatsanraman** [comment]
Build is green. Merging.

**Pull request merged**